File: examples/x3_oop_frog.py
#!/usr/bin/env python3
'''
@author Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

import g2d
from actor import Actor, Arena
from x3_oop_vehicle import Vehicle

class Frog(Actor):
    def __init__(self, arena, pos: (int, int)):
        self._x0, self._y0 = pos
        self._x, self._y = pos
        self._w, self._h = 20, 20
        self._dx, self._dy = 0, 0
        self._speed, self._steps, self._count = 4, 8, 0
        self._arena = arena
        arena.add(self)

    def move(self):
        if self._count > 0:
            self._count -= 1

            arena_w, arena_h = self._arena.size()
            self._y += self._dy
            if self._y < 0:
                self._y = 0
            elif self._y > arena_h - self._h:
                self._y = arena_h - self._h

            self._x += self._dx
            if self._x < 0:
                self._x = 0
            elif self._x > arena_w - self._w:
                self._x = arena_w - self._w

    def position(self):
        return self._x, self._y, self._w, self._h

    def symbol(self):
        return 0, 0

    def collide(self, other):
        self._x, self._y = self._x0, self._y0

    def jump_left(self):
        if self._count == 0:
            self._count = self._steps
            self._dx, self._dy = -self._speed, 0

    def jump_right(self):
        if self._count == 0:
            self._count = self._steps
            self._dx, self._dy = +self._speed, 0

    def jump_up(self):
        if self._count == 0:
            self._count = self._steps
            self._dx, self._dy = 0, -self._speed

    def jump_down(self):
        if self._count == 0:
            self._count = self._steps
            self._dx, self._dy = 0, +self._speed

def tick():
    if g2d.key_pressed("ArrowUp"):
        frog.jump_up()
    elif g2d.key_pressed("ArrowRight"):
        frog.jump_right()
    elif g2d.key_pressed("ArrowDown"):
        frog.jump_down()
    elif g2d.key_pressed("ArrowLeft"):
        frog.jump_left()

    g2d.clear_canvas()
    arena.move_all()
    for a in arena.actors():
        g2d.fill_rect(a.position())

def main():
    global arena, frog
    arena = Arena((480, 360))
    frog = Frog(arena, (230, 340))
    Vehicle(arena, (40, 40), 5)
    Vehicle(arena, (80, 80), -5)
    g2d.init_canvas(arena.size())
    g2d.main_loop(tick)

main()

File: go/x3_oop_frog.go
package main

import . "g2d"

var arena = NewArena(Point{480, 360})
var hero = NewFrog(arena, Point{160, 160})
var a1 = NewVehicle(arena, Point{40, 40}, 5)
var a2 = NewVehicle(arena, Point{120, 40}, 5)
var a3 = NewVehicle(arena, Point{80, 80}, -5)


//// Frog

type Frog struct {
    arena               *Arena
    x, y, w, h          int
    x0, y0, dx, dy      int
    speed, steps, count int
}

func NewFrog(arena *Arena, pos Point) *Frog {
    a := &Frog{arena,
               pos.X, pos.Y, 20, 20,
               pos.X, pos.Y, 0, 0,
               4, 5, 0}
    arena.Add(a)
    return a
}

func (a *Frog) Move() {
    if a.count > 0 {
        a.count--
        as := a.arena.Size()
        a.x += a.dx
        if a.x < 0 {
            a.x = 0
        } else if a.x > as.X-a.w {
            a.x = as.X - a.w
        }
        a.y += a.dy
        if a.y < 0 {
            a.y = 0
        } else if a.y > as.Y-a.h {
            a.y = as.Y - a.h
        }
    }
}

func (a *Frog) JumpLeft() {
    if a.count == 0 {
        a. count = a.steps
        a.dx, a.dy = -a.speed, 0
    }
}

func (a *Frog) JumpRight() {
    if a.count == 0 {
        a. count = a.steps
        a.dx, a.dy = +a.speed, 0
    }
}

func (a *Frog) JumpUp() {
    if a.count == 0 {
        a. count = a.steps
        a.dx, a.dy = 0, -a.speed
    }
}

func (a *Frog) JumpDown() {
    if a.count == 0 {
        a. count = a.steps
        a.dx, a.dy = 0, +a.speed
    }
}

func (a *Frog) Collide(other Actor) {
    a.x = a.x0
    a.y = a.y0
}

func (a *Frog) Position() Point {
    return Point{a.x, a.y}
}

func (a *Frog) Size() Point {
    return Point{a.w, a.h}
}

func (a *Frog) Symbol() Point {
    return Point{0, 20}
}


//// Vehicle

type Vehicle struct {
    x, y, w, h      int
    dx, left, right int
}

func NewVehicle(arena *Arena, pos Point, dx int) *Vehicle {
    a := &Vehicle{pos.X, pos.Y, 20, 20, dx, -100, arena.Size().X+100}
    arena.Add(a)
    return a
}

func (a *Vehicle) Move() {
    if a.x + a.dx < a.left {
        a.x = a.right
    }
    if a.x + a.dx > a.right {
        a.x = a.left
    }
    a.x += a.dx
}

func (a *Vehicle) Position() Point {
    return Point{a.x, a.y}
}

func (a *Vehicle) Size() Point {
    return Point{a.w, a.h}
}

func (a *Vehicle) Symbol() Point {
    return Point{0, 0}
}

func (a *Vehicle) Collide(other Actor) {
}


//// main

func tick() {
    if KeyPressed("ArrowUp") {
        hero.JumpUp()
    } else if KeyPressed("ArrowRight") {
        hero.JumpRight()
    } else if KeyPressed("ArrowDown") {
        hero.JumpDown()
    } else if KeyPressed("ArrowLeft") {
        hero.JumpLeft()
    }

    arena.MoveAll()
    ClearCanvas()
    for _, a := range arena.Actors() {
        FillRect(a.Position(), a.Size())
    }
}

func main() {
    InitCanvas(arena.Size())
    MainLoop(tick)
}
